### PR TITLE
`0.12`: Add docs explaining core features

### LINKED ISSUE #193: [REQUEST] Document how to Build a Database Connector
Explain how to build a Database connector in a guide or other documentation.

### PR DESCRIPTION
**Pull request by zprobst**
Expands documentation to define core reasoning for features and design of `0.12`. 

Also closes #193

### CODE REVIEW

File: docs/technology.md [hunk 1]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
 It takes large chunks of data, converts it all to queries and submits generally long running queries it has to await.
 While this is happening, python can suspend the task waiting on the database allowing the other steps to continue processing.
 This allows for the extractor and interpreter to continue processing data and not be constrained by the database writer.
+
+## Migrations 
+
+Nodestream from versions `0.1`` through `0.10`` automatically requested the creation of indexes and constraints on the database.
```
Comment (zprobst) [inline docs/technology.md:49]:
```suggestion
`nodestream` from versions `0.1`` through `0.10` automatically requested the creation of indexes and constraints on the database. 
```

File: docs/technology.md [hunk 2]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+
+Nodestream from versions `0.1`` through `0.10`` automatically requested the creation of indexes and constraints on the database. 
+This was done by introspecting the schema of the entire project and generating the appropriate queries to create the indexes and constraints.
+This was a very powerful feature but it had a few drawbacks:
+- **It was redundant.** The same indexes and constraints were being created with `IF NOT EXISTS` clauses every time the pipeline was run.
+- **It was slow.** The queries were being executed serially and the pipeline was locked until they were all completed.
+- **It was error prone.** If the database was not in a state that allowed for the creation of the indexes and constraints, the pipeline would fail.
+- **It was high friction.** There was no way to refactor the database without manual intervention. If the schema changed, the pipeline would fail and the user would have to manually remove the indexes, constraints, and sometimes data before running the pipeline again.
+
+To address these issues, nodestream 0.11 introduced the concept of migrations.
```
Comment (zprobst) [inline docs/technology.md:57]:
```suggestion
To address these issues, `nodestream` 0.11 introduced the concept of migrations.
```

File: docs/technology.md [hunk 3]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+
+To achieve these goals, `nodestream` provides a set of commands that allow the developer to manage the migrations.
+
+- `nodestream migratations make [--pipeline <pipeline_name> --scope <scope_name>]` - Diff the current state of the project with the migrations and create a new migration if there are any differences. If `--pipeline` is specified, only migrations for that pipeline will be created. If `--scope` is specified, only migrations for that scope will be created.
+- `nodestream migrations run [--dry-run]` - Apply all migrations to the database. If `--dry-run` is specified, the migrations will not be applied but the list of migrations that would be applied will be printed.
+- `nodestream migrations show -t <target>` - Show the migrations that will be applied to the target database.
+- `nodestream run --auto-migrate` - Run the pipeline and apply any migrations that need to be applied before. 
+
+Additionally, when migrating to nodestream 0.11, the developer can run `nodestream migrations make` to generate the initial set of migrations for the project that will be compatible with 0.10 and below indexes and constraints.
```
Comment (zprobst) [inline docs/technology.md:119]:
```suggestion
Additionally, when migrating to `nodestream` 0.11, the developer can run `nodestream migrations make` to generate the initial set of migrations for the project that will be compatible with 0.10 and below indexes and constraints.
```

File: docs/technology.md [hunk 4]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+- **Creating Node / Property Types.** Registering new node and property types with the database and creating a constraint on the node type.
+- **Renaming Node / Relationship Types** Renaming a node or relationship type.
+- **Deleting Node / Relationship Types** Deleting a node or relationship type.
+- **Adding Properties** Adding a property on a node or relationship type with a default value.
+- **Renaming Properties** Renaming a property on a node or relationship type.
+- **Deleting Properties** Deleting a property on a node or relationship type.
+- **Extending Keys** Adding a property to a key constraint.
+- **Renaming Keys** Renaming a key constraint property. 
+
+Notably, some use cases that are not supported:
```
Comment (zprobst) [inline docs/technology.md:77]: Add a note about rolling back migrations. This is out of scope because we 
1.) aren't losing features because we don't have this today. 
2.) this adds additional complexity and requires further study.

File: docs/technology.md [hunk 5]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+
+The developer experience has several core goals:
+
+- **Low friction.** The developer should be able to make changes to the database schema without having to manually manage the migrations.
+- **Fast.** The developer should be able to make changes to the database schema without having to wait for the migrations to be applied.
+- **Safe.** The developer should be able to make changes to the database schema without having to worry about breaking the pipeline.
+
+To achieve these goals, `nodestream` provides a set of commands that allow the developer to manage the migrations.
+
+- `nodestream migratations make [--pipeline <pipeline_name> --scope <scope_name>]` - Diff the current state of the project with the migrations and create a new migration if there are any differences. If `--pipeline` is specified, only migrations for that pipeline will be created. If `--scope` is specified, only migrations for that scope will be created.
```
Comment (zprobst) [inline docs/technology.md:114]: add a `--dry-run` option to see what the migration would consist of if required.

File: docs/technology.md [hunk 6]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+- `nodestream run --auto-migrate` - Run the pipeline and apply any migrations that need to be applied before. 
+
+Additionally, when migrating to nodestream 0.11, the developer can run `nodestream migrations make` to generate the initial set of migrations for the project that will be compatible with 0.10 and below indexes and constraints.
+
+
+### Connector API
+
+The connector API for migrations defines how the database backend should interface with the migrations framework.
+It is an extensions of the existing `nodestream.databases:DatabaseConnector` class. 
+Starting with `0.11`, the `DatabaseConnector` class will be extended with the following optional method:
```
Comment (zprobst) [inline docs/technology.md:126]:
```suggestion
Starting with `0.12`, the `DatabaseConnector` class will be extended with the following optional method:
```

File: docs/technology.md [hunk 7]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+
+`Nodestream` is designed to be extensible. 
+The core of `nodestream` is designed to be as generic as possible and provide a set of extension points that allow for the creation of plugins that can be used to extend the functionality of nodestream.
+
+One of the most important extension points is the `DatabaseConnector` class.
+This class is responsible for interfacing with the database and providing a consistent interface for the rest of the application to interact with the database.
+The `DatabaseConnector` class is an abstract class that defines the interface that all database connectors must implement.
+While most projects won't need to implement their own database connector, every project will need to use one. 
+
+Prior to nodestream 0.11, `neo4j` was housed in the core of nodestream.
```
Comment (zprobst) [inline docs/technology.md:157]:
```suggestion
Prior to nodestream 0.12, `neo4j` was housed in the core of nodestream. 
```

File: docs/technology.md [hunk 8]
```diff
@@ -43,3 +43,119 @@ The slowest step in the pipeline is the graph database writer.
...
+
+One of the most important extension points is the `DatabaseConnector` class.
+This class is responsible for interfacing with the database and providing a consistent interface for the rest of the application to interact with the database.
+The `DatabaseConnector` class is an abstract class that defines the interface that all database connectors must implement.
+While most projects won't need to implement their own database connector, every project will need to use one. 
+
+Prior to nodestream 0.11, `neo4j` was housed in the core of nodestream. 
+This meant that every project that used nodestream had to install `neo4j` even if they didn't use it.
+This was not ideal because it added additional dependencies to the project and increased the size of a final application.
+To address this, `nodestream` 0.11 pulled `neo4j` out of the core and into a separate package called `nodestream-neo4j`.
```
Comment (zprobst) [inline docs/technology.md:160]:
```suggestion
To address this, `nodestream` 0.12 pulled `neo4j` out of the core and into a separate package called `nodestream-neo4j`.
```

**Review by grantleehoffman — APPROVED** [review]

File: docs/docs/guides/creating-your-own-database-connector.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
+# Creating A Database Connector
+
+## Introduction
+
+This guide will walk you through the process of creating a database connector for nodestream.
```
Comment (yasonk) [inline docs/docs/guides/creating-your-own-database-connector.md:5]: The start of the two sentences is very similar. Maybe the second sentence can start like 
```
We will create a database...
```

Or use a bullet list maybe?

File: docs/docs/guides/creating-your-own-database-connector.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
+# Creating A Database Connector
+
+## Introduction
+
+This guide will walk you through the process of creating a database connector for nodestream. 
+This guide will walk through the process of creating a database connector for a hpypothetical graph database called `smith`.
```
Comment (yasonk) [inline docs/docs/guides/creating-your-own-database-connector.md:6]: typo in `hypothetical`

File: docs/docs/guides/creating-your-own-database-connector.md [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    Also it is likely that you will not need to build a database connector unless you are using a graph database that is not supported by nodestream.
+
+
+## Prerequisites
+
+Before you begin, you will need to have a basic understanding of the following concepts:
+
+- **Labeled Property Graphs** The graph database that you are creating a connector for should be a [labeled property graph](https://www.oxfordsemantic.tech/faqs/what-is-a-labeled-property-graph).
+- **Cypher** Ideally, the graph database that you are creating a connector for should have a query language similar to [Cypher](https://neo4j.com/developer/cypher-query-language/).
+- **Python** You should have an intermidiate understanding of Python.
```
Comment (yasonk) [inline docs/docs/guides/creating-your-own-database-connector.md:20]: typo in word `Intermediate`

File: docs/technology.md [hunk 12]
```diff
@@ -43,3 +43,121 @@ The slowest step in the pipeline is the graph database writer.
 It takes large chunks of data, converts it all to queries and submits generally long running queries it has to await.
 While this is happening, python can suspend the task waiting on the database allowing the other steps to continue processing.
 This allows for the extractor and interpreter to continue processing data and not be constrained by the database writer.
+
+## Migrations
```
Comment (yasonk) [inline docs/technology.md:47]: Not in the changes, but above `asyncronously` should be `asynchronously` and also
`exectuion`
and also
`amout`

Sorry I asked chatGPT to check the entire file :)

File: docs/technology.md [hunk 13]
```diff
@@ -43,3 +43,121 @@ The slowest step in the pipeline is the graph database writer.
...
+- **It was high friction.** There was no way to refactor the database without manual intervention. If the schema changed, the pipeline would fail and the user would have to manually remove the indexes, constraints, and sometimes data before running the pipeline again.
+
+To address these issues, `nodestream` 0.12 introduced the concept of migrations.
+Migrations are a way of encapsulating changes to the database schema in a way that can be applied incrementally. 
+Conceptually, they are similar to the migrations in the [Django](https://docs.djangoproject.com/en/5.0/topics/migrations/), [Rails](https://guides.rubyonrails.org/v3.2/migrations.html), [Neo4j Migrations](https://neo4j.com/labs/neo4j-migrations/2.0/), and [Flyway](https://documentation.red-gate.com/fd/migrations-184127470.html) frameworks.
+
+
+### Scope
+
+Some changes to the database are tricier to deal with than others, for the 0.11 release, the following changes are supported:
```
Comment (yasonk) [inline docs/technology.md:64]: `tricier` chatGPT did not catch this :(

File: docs/technology.md [hunk 14]
```diff
@@ -43,3 +43,121 @@ The slowest step in the pipeline is the graph database writer.
...
+- **Creating indexes.** Creating indexes on node/relationship properties.
+- **Droping indexes.** Dropping indexes on node/relationship properties.
```
Comment (yasonk) [inline docs/technology.md:66]: `Droping`

File: docs/technology.md [hunk 15]
```diff
@@ -43,3 +43,121 @@ The slowest step in the pipeline is the graph database writer.
...
+
+The developer experience has several core goals:
+
+- **Low friction.** The developer should be able to make changes to the database schema without having to manually manage the migrations.
+- **Fast.** The developer should be able to make changes to the database schema without having to wait for the migrations to be applied.
+- **Safe.** The developer should be able to make changes to the database schema without having to worry about breaking the pipeline.
+
+To achieve these goals, `nodestream` provides a set of commands that allow the developer to manage the migrations.
+
+- `nodestream migratations make [--pipeline <pipeline_name> --scope <scope_name>]` - Diff the current state of the project with the migrations and create a new migration if there are any differences. If `--pipeline` is specified, only migrations for that pipeline will be created. If `--scope` is specified, only migrations for that scope will be created.
```
Comment (yasonk) [inline docs/technology.md:116]: `migratations`

File: docs/technology.md [hunk 16]
```diff
@@ -43,3 +43,121 @@ The slowest step in the pipeline is the graph database writer.
...
+    # class contents omitted for brevity 
+    @abstractmethod
+    def get_migrator(self) -> Migrator:
+        pass
+```
+
+The `Migrator` class is responsible for executing the migrations on the database.
+It's interface is defined as follows: 
+
+- `Migrator.exectute_operation(operation: Operation) -> None` - Execute the given operation on the database. This method handles the logic of executing the operation and raising an exception if the operation fails.
```
Comment (yasonk) [inline docs/technology.md:143]: `exectute`

File: docs/technology.md (hunk 12 above)
Comment (zprobst) [reply docs/technology.md:47]: Thank you!!

**Review by yasonk — APPROVED** [review]
